### PR TITLE
docs: add model debug doc

### PR DESCRIPTION
**Pull request by rb-determined-ai**
## Description

It is looking like the observability project is going to solve what I was trying to solve in https://github.com/determined-ai/determined/pull/1625, only much better.  As a result, I'm pulling out just the debug documentation as a separate PR to land now.

### CODE REVIEW

File: docs/how-to/model-debug.txt [hunk 1]
```diff
@@ -0,0 +1,337 @@
...
+Step 2.
+
+Probably you already know your code works, but this is a reminder to
+double-check before proceeding.
+
+2. Does each method of your Trial class work locally?
+-----------------------------------------------------
+
+This step assumes you have a working local environment for training. If
+you normally run your code in a docker environment of some instead, skip
```
Comment (neilconway) [inline docs/how-to/model-debug.txt:53]: Grammar ("some instead"?)

File: docs/how-to/model-debug.txt [hunk 2]
```diff
@@ -0,0 +1,337 @@
...
+you do not, skip to Step 4.
+
+**How to test:** Simply create an experiment, passing ``--local --test``
+on the command line:
+
+.. code:: bash
+
+   det experiment create myconfig.yaml my_model_dir --local --test
+
+**How to diagnose failures:** Local test mode does very few things; it
```
Comment (neilconway) [inline docs/how-to/model-debug.txt:83]: Non-blocking: I might move the explanation of what local test mode does up a bit higher, to the rationale for why you should do this debugging step.

I also feel like this discussion might be improved by a concrete example -- e.g., what kind of problem might be caught by local test mode that wouldn't be caught by the earlier debugging steps?

File: docs/how-to/model-debug.txt [hunk 3]
```diff
@@ -0,0 +1,337 @@
...
+      experiment config.
+
+   -  Any ``pip``-install or ``apt``-install commands needed in the
+      interactive environment must either be built into a custom
+      docker image or they can be written into file called
+      ``startup-hook.sh`` in the root of the model definition
+      directory.  The commands in ``startup-hook.sh`` will be executed
+      automatically before training begins.
+
+      The startup-hook.sh option is quiker to try but will be slower at
```
Comment (neilconway) [inline docs/how-to/model-debug.txt:224]: Typo.

File: docs/how-to/model-debug.txt [hunk 4]
```diff
@@ -0,0 +1,337 @@
...
+   det experiment create myconfig.yaml my_model_dir -f
+
+**How to diagnose failures:**  If you are using the ``determined``
+library APIs correctly, then theoretically distributed training should
+"just work".  However, you should be aware of some common pitfalls:
+
+   - Determined normally controls the details of distributed training
+     for you.  If you are also trying to control those details, such as
+     calling ``tf.config.set_visible_devices()`` in a KerasTrial or
+     EstimatorTrial will very likely cause issues.
```
Comment (neilconway) [inline docs/how-to/model-debug.txt:328]: Grammar ("If you are trying to " -> "Attempting to", for example).

File: docs/how-to/model-debug.txt [hunk 5]
```diff
@@ -0,0 +1,337 @@
...
+
+Then create your experiment:
+
+.. code:: bash
+
+   det experiment create myconfig.yaml my_model_dir -f
+
+**How to diagnose failures:**  If you are using the ``determined``
+library APIs correctly, then theoretically distributed training should
+"just work".  However, you should be aware of some common pitfalls:
```
Comment (neilconway) [inline docs/how-to/model-debug.txt:323]: Talk about networking issues? And maybe scheduling problems / hangs?

**Review by neilconway — COMMENTED** [review]
Overall, I think this is great! I added a few minor comments.

I think the discussion might benefit from a bit more elaboration on the thought process behind these debugging steps. For example, you could start by saying "code running in Determined is different from code running in your local environment in several ways: (a) runs as part of the Trial API (b) runs in a container on a remote machine (c) might run as part of distributed training. When debugging, it helps to isolate whether any of those environmental differences are the root cause of any observed issues." And then add some details to each section, "now that we have confirmed that the code works locally, the next step is to ..."

File: docs/how-to/model-debug.txt (hunk 5 above)
Comment (rb-determined-ai) [reply docs/how-to/model-debug.txt:323]: Added a detail about not being scheduled.

Are there common networking issue that affect distributed training and not normal training?   Almost all of the networking issues I've helped debug were issues that affected even a single notebook.

re hangs: I think hangs in our harness are nearly 100% our responsibility.  I can't think of any hangs that were due to the user doing something wrong.

Comment (neilconway) [reply docs/how-to/model-debug.txt:323]: Yeah -- I mostly meant hangs due to dtrain trials not being scheduled.

Re: networking, I was thinking of issues like: (a) matching network interface names (b) additional ports we use for NCCL, ssh, etc. traffic we don't use for single-GPU training (c) issues with istio or other proxies that seem to be more problematic for dtrain trials (e.g., https://determined-community.slack.com/archives/CV3MTNZ6U/p1607112511338900)

File: docs/how-to/model-debug.txt (hunk 2 above)
Comment (rb-determined-ai) [reply docs/how-to/model-debug.txt:112]: Rewrote steps 1 and 2, let me know what you think.

File: docs/how-to/model-debug.txt (hunk 5 above)
Comment (rb-determined-ai) [reply docs/how-to/model-debug.txt]: I think things like port configurations belong in a separate "debug/validate your determined cluster", where we would have a series of tests that new cluster admins could run to ensure that their users' code would work.  I think this document should target the ml end user

**Comment by rb-determined-ai** [comment]
Danny is going to update the `rstfmt` command to support non-auto-enumerated lists so I can pass `make -C doc check`, otherwise I got rid of all TODOs in the doc and I think it's ready for another round of review.

File: docs/how-to/model-debug.txt (hunk 2 above)
Comment (neilconway) [reply docs/how-to/model-debug.txt:113]: sg.

**Review by neilconway — APPROVED** [review]
lgtm. Can we find a few places to add links to this doc elsewhere in the docs? e.g., maybe add it as a suggestion in the Keras/PyTorch tutorials, or add an FAQ, etc.

**Comment by rb-determined-ai** [comment]
I added it in several places:
  - FAQ
  - Keras/Pytorch/Estimator API docs
  - Model definition topic guides

**Pull request merged**